src/d3a/d3a_core/sim_results/export_unmatched_loads.py

```diff
@@ -15,127 +15,160 @@
+        indict[area.name] = {}
+        for child in area.children:
+            self.name_uuid_map[child.name] = child.uuid
+            if child.children:
+                indict[area.name] = self.find_unmatched_loads(child, indict[area.name])
+            else:
+                if isinstance(child.strategy, LoadHoursStrategy):
+                    indict[area.name][child.name] = self._accumulate_unmatched_loads(child)
+        return indict
+
+    @classmethod
+    def _accumulate_unmatched_loads(cls, area):
+        """
+        actually determines the unmatched loads
+        """
+        unmatched_times = []
+        for market in area.parent.past_markets:
+            desired_energy_Wh = area.strategy.state.desired_energy_Wh[market.time_slot]
+            traded_energy_kWh = market.traded_energy[area.name] \
+                if market is not None and (area.name in market.traded_energy) \
+                else 0.0
+            deficit = desired_energy_Wh + traded_energy_kWh * 1000.0
+            if deficit > FLOATING_POINT_TOLERANCE:
+                unmatched_times.append(market.time_slot)
+        return {"unmatched_times": unmatched_times}
+
+    def change_name_to_uuid(self, indict):
+        """
+        postprocessing: changing area names to uuids
+        """
+        new = {}
+        for k, v in indict.items():
+            new[self.name_uuid_map[k]] = v
+        return new
+
+    def _accumulate_all_uls_in_branch(self, subdict, unmatched_list) -> list:
+        """
+        aggregate all UL times in one branch into one list
+        """
+        for key in subdict.keys():
+            if key == "unmatched_times":
+                return unmatched_list + subdict[key]
+            else:
+                unmatched_list = self._accumulate_all_uls_in_branch(subdict[key], unmatched_list)
+        return unmatched_list
+
+    def expand_ul_to_parents(self, subdict, parent_name, outdict):
+        """
+        expand UL times to all nodes including all UL times of the sub branch
+        """
+
+        for node_name, subsubdict in subdict.items():
+            if isinstance(subsubdict, dict):
+                # nodes
+                ul_list = self._accumulate_all_uls_in_branch(subsubdict, [])
+                sub_ul = sorted(list(set(ul_list)))
+                if parent_name in outdict:
+                    outdict[parent_name].update({node_name: sub_ul})
+                else:
+                    outdict[parent_name] = {node_name: sub_ul}
+                outdict.update(self.expand_ul_to_parents(subsubdict, node_name, {}))
+            else:
+                # leafs
+                outdict[parent_name] = {parent_name: subsubdict}
+
+        return outdict
+
+    @classmethod
+    def _get_hover_info(cls, indict, hour_time):
+        """
+        returns dict of UL for each subarea for the hover the UL graph
+        """
+        hover_dict = {}
+        ul_count = 0
+        for child_name, child_ul_list in indict.items():
```

Comment:
I was wondering whether it is possible to use groupby to perform this grouping [like here](https://stackoverflow.com/a/25137532). No need to try it out, I was just wondering if it would be easier that way.

Reply:
Sorry, I dont see how this could be helpful here. Maybe you could explain offline.